### DISCUSSION TITLE
while loops, invariants, break statements

### DISCUSSION THREAD

**Discussion by tjhance** [verus-lang/verus#108] [Language design]
Currently, `while` loops allow "complex conditions", i.e., conditions that are not pure expressions. Right now, the conversion to AIR looks something like this:

```rust
// Verus:
while { cond_stmts; cond_expr }
{
    invariant(inv);

    body;
}
after_loop;
```

```
// AIR:
assert(inv);
new context { // loop body
    assume(inv);
    cond_stmts;
    assume(cond_expr);
    body;
    assert(inv);
}
new context { // after the loop
    assume(inv);
    cond_stmts;
    assume(!cond_expr);
    after_loop;
}
```

Note also that `cond_stmts` may have asserts or preconditions in it, which need to follow from the `inv`. (Minor note: I think we're being a little inefficient right now proving those twice, though that's probably an easy fix.)

We can see that this make sense in a `while` loop with no `break` statements, because the last thing to happen in the loop _must_ be that the loop condition was evaluated and the result was `false`.

Unfortunately, things get a bit more difficult when we handle `break` statements because:

 1. we can no longer make the claim that `cond_stmts` were the last things to execute, and
 2. we can't claim that `cond_expr` evaluated to `false`.

(BTW, point (2) is kind of an issue in Dafny too. It's not a _huge_ deal, because Dafny's control flow analysis seems to do the right thing in most cases, but this is fairly opaque to the user, and I hope we can come up with something cleaner. Also, since Verus supports complex conditions, point (1) exacerabes the problem for us.)

Here are a handful of proposals:

### Don't support complex conditions.

Self-explanatory.

Pros: makes things easier on Verus.

Cons: makes things harder for users.

### If the loop contains a `break` statement, just "optionally" perform the cond_stmts:

```
new context { // after the loop
    assume(inv);
    if (nondeterministic) {
        cond_stmts;
        assume(!cond_expr);
    } else {
    }
    after_loop;
}
```

Pro: this is probably the smallest diff to the system we have in place now.

Con: I think this would probably be difficult to work with. Even if we ignored the `cond_stmts`, there's still the fact that we can't rely on `!cond_expr` to hold. (See my note about Dafny, above.)

### Have specific `on_exit` invariants, and use those instead of the `invariants`

Something like this:

```rust
// Verus:
while { cond_stmts; cond_expr }
{
    invariant(inv);
    on_exit(exit_inv);

    body;
}
after_loop;
```

```
// AIR:
assert(inv);
new context { // loop body
    assume(inv);
    cond_stmts;
    assume(cond_expr);

    body {
        // ...
        
        // at each continue:
        assert(inv);

        // at each break:
        assert(exit_inv); // not asserting `inv`
    }

    assert(inv);
}
new context { // where we evaluate the condition and it fails
    assume(inv);
    cond_stmts;
    assume(!cond_expr);
    assert(exit_inv);
}
new context { // after the loop
    assume(exit_inv); // only assuming `exit_inv`, not `inv`
    after_loop;
}
```

In other words, the user has to specify exactly what they want to hold after the loop exits using explicit `on_exit` predicates.

I already think it's a good idea to have `on_exit` predicates (see the brainstorm discussion: https://github.com/secure-foundations/verus/discussions/14). However, it would probably be a little annoying that the user doesn't also get the benefits of the `invariant`s after the loop exits.

Also note that we would have to either:

 * Apply this rule _all_ the time (which hurts the experience for simple loops)
 * Apply this rule only for complex rules (which makes the rules hard for users to understand)

### Move the 'point where the invariant is true' to be _after_ the evaluation of the `cond_stmts`

Something like this:

```rust
// Verus:
while { cond_stmts; cond_expr }
{
    invariant(inv);
    on_exit(exit_inv);

    body;
}
after_loop;
```

```
// AIR:

cond_stmts;
assert(inv);
assert(!cond_expr >>= exit_inv);

new context { // loop body
    cond_stmts; // change any asserts to assumes
    assume(inv);
    assume(cond_expr);

    body {
        // ...
        
        // at each continue:
        cond_stmts;
        assert(inv);
        assert(!cond_expr >>= exit_inv);

        // at each break:
        assert(inv);
        assert(exit_inv);
    }

    cond_stmts;
    assert(inv);
    assert(!cond_expr >>= exit_inv);
}

new context { // after the loop
    assume(inv);
    assume(exit_inv);
    if there are no break statements {
        assume(!cond_expr);
    }
    after_loop;
}
```

Note that `cond_stmts` now appears in a bunch of places (on loop entry; at the end of the loop body; on any continue statement). So anything in `cond_stmts` has to be proved at each point, although this isn't so different from having to prove the invariants at each point. Also, we can annotate the error messages for any `assert` statements in the `cond_stmts` to explain "where" the error is taking place.

(One interesting property of this proposal is that the conditions resemble what the user would have to manually do as a workaround if we didn't support `cond_stmts` at all. For example: suppose the user wanted to put `complex_method_call()` in the loop condition. If Verus doesn't support `cond_stmts`, they would instead have to put the `complex_method_call()` in a bunch of different places: before the loop, at the end of the body, and at each `continue` statement. That is, the same place that we put `cond_stmts` in this scenario.) 

This also has the advantage that the `invariant` statement in the user's verus code is actually sitting at the point where the predicate is supposed to hold. Note that this wasn't true in any of the other scenarios. For example, this currently passes in Verus, and it would pass in any of the other scenarios:

```rust
fn main() {
    let mut b = 5;

    while ({
                           //   <-----------------------------------------------
      b = 6;               //                                                  |
      true                 //                                                  |
    }) {                   //                                                  |
        invariant(b == 5); // this passes because Verus proves the invariant here
        assert(b == 6);
        b = 5;
    }
}
```

Whereas in this last scenario, you would do:

```rust
fn main() {
    let mut b = 5;

    while ({
      b = 6;      
      true 
    }) {
        invariant(b == 6);
        assert(b == 6);
        b = 5;
    }
}
```

**Comment by tjhance** [comment]
I have thought about this some more, and I think all these approaches I listed are coming from the wrong angle.

Until now, I (and we) have been thinking in terms of the textbook definition of a loop invariant, that is, “a loop invariant is a predicate that holds at the beginning and end of each iteration, including on entrance and exit to the loop”. And all the above proposals grappled with the question of “where is the boundary between iterations?”

However, (as @jaybosamiya pointed out in the other discussion) we could imagine invariants being _anywhere_ in a loop (or at least, to within our reasonable ability to generate VCs for it).

If we start from this angle, it frees up our thinking quite a bit. Instead of determining a specific place for the invariants to go, we could just put them at any program point, and say that point is where they apply.

For example:

```rust
    while ({
      invariant_here(/* invariant could go here */);
      let b = complex_condition();      
      b 
    }) {
        invariant_here(/* or here if it suits the user better */);
        b = 5;
        invariant_here(/* or even here */);
    }
    invariant_on_exit(/* exit invariants should go here, at the exit, which is where they're supposed to hold */);
```

When we look at it this way, the distinction between the loop "condition" and the loop "body" now looks like a red herring. The loop starts at the beginning of the condition, and the loop exit condition looks like any other `break` statement.

Now, we _do_ know from experience that we frequently _do_ want the “textbook invariants“ that hold both on entry and exit. Unfortunately, it isn't possible to find a single program point where such invariants should go. We could have a special convenience mechanism for this situation (e.g., a special `invariant` statement that both places an invariant at the point it's at _and_ declares the same predicate as an exit invariant). But internally, we could just operate on `invariant_here` and `invariant_on_exit` statements like in the example above.

**Comment by utaal** [comment]
I haven't looked into this in detail, but it seems to at least partly overlap with #14 . If so, could you please merge the two and mark the other one as resolved? Thanks!

**Reply by tjhance** [reply]
I think this is a pretty different discussion - the earlier one was meant to be a feature-brainstorming hub that happened to have a few while loop related features in it.
